src/Features/Core/Portable/DisposeAnalysis/DisposeObjectsBeforeLosingScopeDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,202 @@
+using Microsoft.CodeAnalysis.FlowAnalysis;
+using Microsoft.CodeAnalysis.FlowAnalysis.DataFlow;
+using Microsoft.CodeAnalysis.FlowAnalysis.DataFlow.DisposeAnalysis;
+using Microsoft.CodeAnalysis.FlowAnalysis.DataFlow.PointsToAnalysis;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+
+namespace Microsoft.CodeAnalysis.DisposeAnalysis
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp, LanguageNames.VisualBasic)]
+    internal sealed class DisposeObjectsBeforeLosingScopeDiagnosticAnalyzer
+        : AbstractCodeQualityDiagnosticAnalyzer
+    {
+        private static readonly DiagnosticDescriptor s_disposeObjectsBeforeLosingScopeRule = CreateDescriptor(
+            IDEDiagnosticIds.DisposeObjectsBeforeLosingScopeDiagnosticId,
+            title: new LocalizableResourceString(nameof(FeaturesResources.Dispose_objects_before_losing_scope), FeaturesResources.ResourceManager, typeof(FeaturesResources)),
+            messageFormat: new LocalizableResourceString(nameof(FeaturesResources.Disposable_object_created_by_0_is_never_disposed), FeaturesResources.ResourceManager, typeof(FeaturesResources)),
+            description: new LocalizableResourceString(nameof(FeaturesResources.UseRecommendedDisposePatternDescription), FeaturesResources.ResourceManager, typeof(FeaturesResources)),
+            isUnneccessary: false);
+
+        private static readonly DiagnosticDescriptor s_useRecommendedDisposePatternRule = CreateDescriptor(
+            IDEDiagnosticIds.UseRecommendedDisposePatternDiagnosticId,
+            title: new LocalizableResourceString(nameof(FeaturesResources.Use_recommended_dispose_pattern), FeaturesResources.ResourceManager, typeof(FeaturesResources)),
+            messageFormat: new LocalizableResourceString(nameof(FeaturesResources.Use_recommended_dispose_pattern_to_ensure_that_object_created_by_0_is_disposed_on_all_paths_using_statement_declaration_or_try_finally), FeaturesResources.ResourceManager, typeof(FeaturesResources)),
+            description: new LocalizableResourceString(nameof(FeaturesResources.UseRecommendedDisposePatternDescription), FeaturesResources.ResourceManager, typeof(FeaturesResources)),
+            isUnneccessary: false);
+
+        public DisposeObjectsBeforeLosingScopeDiagnosticAnalyzer()
+            : base(ImmutableArray.Create(s_disposeObjectsBeforeLosingScopeRule, s_useRecommendedDisposePatternRule), GeneratedCodeAnalysisFlags.None)
+        {
+        }
+
+        public override bool OpenFileOnly(Workspace workspace) => false;
+
+        public override DiagnosticAnalyzerCategory GetAnalyzerCategory() => DiagnosticAnalyzerCategory.SemanticDocumentAnalysis;
+
+        protected override void InitializeWorker(AnalysisContext context)
+        {
+            context.RegisterCompilationStartAction(compilationContext =>
+            {
+                if (!DisposeAnalysisHelper.TryGetOrCreate(compilationContext.Compilation, out DisposeAnalysisHelper disposeAnalysisHelper))
```

Comment:
so I guess this cache is tied to Compilation in CompilationWithAnalyzer lifetime?

Reply:
It's ConditionalWeakTable right now, see [here](https://github.com/dotnet/roslyn-analyzers/blob/0bb8197ff0551722a507027d8e3c7a97a5e62a9e/src/Utilities/FlowAnalysis/FlowAnalysis/Analysis/DisposeAnalysis/DisposeAnalysisHelper.cs#L32). But good suggestion to actually instead save it per compilation using AnalysisContext.TryGetValue method. I will update the library.

Reply:
I will leave this as is as we do not have any helper methods on AnalysisContext or CompilationStartAnalysisContext to save compilation specific data.